syncstorage/storage/spanner.py

```diff
@@ -0,0 +1,988 @@
+            getq(queries.DELETE_ITEM),
+            params={"userid": userid,
+                    "collectionid": collectionid,
+                    "item": item,
+                    "ttl": ttl},
+            param_types={"userid": param_types.STRING,
+                         "collectionid": param_types.INT64,
+                         "item": param_types.STRING,
+                         "ttl": param_types.TIMESTAMP}
+        )
+        if not rowcount:
+            raise ItemNotFoundError
+        return session.timestamp
+
+    #
+    # Private methods for manipulating collections.
+    #
+
+    def _get_collection_id(self, collection, create=False):
+        """Returns a collection id, given the name.
+
+        If the named collection does not exist then CollectionNotFoundError
+        will be raised.  To automatically create collections on demand, pass
+        create=True.
+
+        These queries run outside any transactions.
+
+        """
+        # Grab it from the cache if we can.
+        try:
+            return self._collections_by_name[collection]
+        except KeyError:
+            pass
+
+        # Try to look it up in the database.
+        with self._database.snapshot() as snapshot:
+            result = snapshot.execute_sql(
+                getq(queries.COLLECTION_ID),
+                params={"name": collection},
+                param_types={"name": param_types.STRING}
+            ).one_or_none()
+
+        if not result:
+            # Shall we auto-create it?
+            if not create:
+                raise CollectionNotFoundError
+
+            # Insert it into the database.  This might raise a conflict
+            # if it was inserted concurrently by someone else.
+            try:
+                # We don't have autoincrement primary keys, so we must read
+                # first to
+                # see if we can try and insert a higher ordered one
+                def create_collection(transaction):
+                    result = transaction.execute_sql(
+                        "SELECT MAX(collectionid) from collections"
+                    ).one()
+                    collectionid = result[0] + 1 if result[0] else 1
+                    # Choose a collection ID at least as large as the first
+                    # custom id
+                    if self.standard_collections:
+                        collectionid = max(collectionid,
+                                           FIRST_CUSTOM_COLLECTION_ID)
+                    transaction.insert(
+                        "collections",
+                        columns=["collectionid", "name"],
+                        values=[[collectionid, collection]]
+                    )
+                    return collectionid
+
+                # Run the transaction and get the collectionid we created
+                collectionid = self._database.run_in_transaction(
+                    create_collection)
+            except Exception as e:
```

Comment:
Pretty broad `except` clause here; will it always be a particular type of exception e.g. `Aborted` like you use above for the locking stuff?